### PR TITLE
RPC for AMM smart Routing

### PR DESCRIPTION
**Pull request by c02fee**
for reference we are implementing a similar router as this: https://github.com/Uniswap/smart-order-router/blob/main/src/routers/alpha-router/alpha-router.ts

### CODE REVIEW

File: pallets/router/rpc/runtime-api/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::Codec;
+use primitives::{Liquidity, Shortfall};
+use sp_runtime::DispatchError;
+
+sp_api::decl_runtime_apis! {
+    pub trait RouterApi<AccountId> where
+        AccountId: Codec, {
+        fn get_smart_router(account: AccountId) -> Result<(Liquidity, Shortfall), DispatchError>;
```
Comment (drbh) [inline pallets/router/rpc/runtime-api/src/lib.rs:24]: this function should take in an input and output token maybe something like 

```rust
get_smart_route(
    token_in: CurrencyId,
    token_out: CurrencyId,
) -> Vec<CurrencyId>
```

and maybe we should rename it to `get_best_route`

File: pallets/router/rpc/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+impl<C, B> Loans<C, B> {
+    /// Create new `Loans` with the given reference to the client.
+    pub fn new(client: Arc<C>) -> Self {
+        Self {
+            client,
+            _marker: Default::default(),
+        }
+    }
+}
```
Comment (drbh) [inline pallets/router/rpc/src/lib.rs:51]: should be renamed

File: pallets/router/rpc/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+pub trait RouterApi<BlockHash, AccountId> {
+    #[rpc(name = "amm_getSmartRouter")]
+    fn get_smart_router(
+        &self,
+        account: AccountId,
+        at: Option<BlockHash>,
+    ) -> Result<(Liquidity, Shortfall)>;
+}
+
+/// A struct that implements the [`LoansApi`].
```
Comment (drbh) [inline pallets/router/rpc/src/lib.rs:37]: should be renamed

**Comment by c02fee** [comment]
@drbh  PR updated

**Comment by c02fee** [comment]
Clippy errors
```
error[E0432]: unresolved import `scale_info::prelude::collections::HashMap`
    --> /home/runner/work/parallel/parallel/pallets/router/src/lib.rs:47:9
     |
  47 |     use scale_info::prelude::collections::HashMap;
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `HashMap` in `collections`

  error: cannot find macro `vec` in this scope
     --> /home/runner/work/parallel/parallel/pallets/router/src/lib.rs:191:38
      |
  191 |                 let mut difference = vec![];
      |                                      ^^^
      |
      = note: consider importing one of these items:
              frame_benchmarking::vec
              scale_info::prelude::vec
              sp_std::vec

  error: cannot find macro `vec` in this scope
     --> /home/runner/work/parallel/parallel/pallets/router/src/lib.rs:172:56
      |
  172 |             let mut queue: Vec<(u32, u32, Vec<u32>)> = vec![(start, end, path)];
      |                                                        ^^^
```

**Comment by c02fee** [comment]
:tada:

**Review by drbh — APPROVED** [review]
lgtm

**Pull request merged**